src/pydicom/pixels/decoders/base.py

```diff
@@ -0,0 +1,1911 @@
+              ``1`` for signed.
+
+            The following options may be used with any transfer syntax:
+
+            * `as_rgb`: :class:`bool` - if ``True`` (default) then convert pixel
+              data with a YCbCr :ref:`photometric interpretation
+              <glossary_photometric_interpretation>` such as ``"YBR_FULL_422"``
+              to RGB.
+            * `force_rgb`: :class:`bool` - if ``True`` then force a YCbCr to RGB
+              color space conversion on the array (default ``False``).
+            * `force_ybr`: :class:`bool` - if ``True`` then force an RGB to YCbCr
+              color space conversion on the array (default ``False``).
+
+            The following options may be used with native (uncompressed)
+            transfer syntaxes:
+
+            * `view_only`: :class:`bool` - if ``True`` then make a best effort
+              attempt to yield an :class:`~numpy.ndarray` that's a `view
+              <https://numpy.org/doc/stable/user/basics.copies.html#view>`_
+              on the original buffer (default ``False``).
+
+            Options for the decoding plugin(s) may also be supplied. See the
+            :doc:`decoding plugin options </guides/decoding/decoder_plugin_options>`
+            for more information.
+
+        Yields
+        ------
+        numpy.ndarray
+            The decoded and reshaped pixel data, with shape:
+
+            * (rows, columns) for single frame, single plane data
+            * (rows, columns, planes) for single frame, multi-plane data
+
+            A writeable :class:`~numpy.ndarray` is returned by default. For
+            native transfer syntaxes with ``view_only=True`` a read-only
+            :class:`~numpy.ndarray` will be yielded if `src` is immutable.
+        """
+        if not HAVE_NP:
+            raise ImportError(
+                "NumPy is required when converting pixel data to an ndarray"
+            )
+
+        runner = DecodeRunner(self.UID)
+        runner.set_source(src)
+        runner.set_options(**kwargs)
+        runner.set_decoders(self._validate_decoders(decoding_plugin))
+
+        if config.debugging:
+            LOGGER.debug(runner)
+
+        if validate:
+            runner.validate_options()
+            runner.validate_buffer()
+
+        if self.is_native:
+            func = self._as_array_native
+            as_writeable = not runner.get_option("view_only", False)
+        else:
+            func = self._as_array_encapsulated
+            as_writeable = True
+
+        if self.is_encapsulated and not indices:
+            for frame in runner.iter_decode():
+                arr = np.frombuffer(frame, dtype=runner.pixel_dtype)
+                arr = runner.reshape(arr, as_frame=True)
+                if raw:
+                    yield arr if arr.flags.writeable else arr.copy()
+                    continue
+
+                arr = runner.process(arr)
+
+                yield arr if arr.flags.writeable else arr.copy()
```

Comment:
Will yield frames from the source data as long as there are frames remaining (and no Basic or Extended Offset Tables), related to #1666